README
NAME

    Plack::Middleware::Proxy::Connect::IO - CONNECT method

SYNOPSIS

        # In app.psgi
        use Plack::Builder;
        use Plack::App::Proxy;
    
        builder {
            enable "Proxy::Connect::IO";
            enable "Proxy::Requests";
            Plack::App::Proxy->new->to_app;
        };

DESCRIPTION

    This middleware handles the CONNECT method. It allows to connect to
    https addresses.

    The middleware runs on servers supporting psgix.io and provides own
    event loop so does not work correctly with psgi.nonblocking servers.

    The middleware uses only Perl's core modules: IO::Socket::INET and
    IO::Select.

SEE ALSO

    Plack, Plack::App::Proxy, Plack::Middleware::Proxy::Connect.

BUGS

    If you find the bug or want to implement new features, please report it
    at
    https://github.com/dex4er/perl-Plack-Middleware-Proxy-Connect-IO/issues

    The code repository is available at
    http://github.com/dex4er/perl-Plack-Middleware-Proxy-Connect-IO

AUTHOR

    Piotr Roszatycki <dexter@cpan.org>

LICENSE

    Copyright (c) 2014, 2016, 2023 Piotr Roszatycki <dexter@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as perl itself.

    See http://dev.perl.org/licenses/artistic.html



Changes
0.0304  2023-06-17      Piotr Roszatycki <dexter@cpan.org>

* Use cpanfile for Module::Build.

0.0303  2023-06-11      Piotr Roszatycki <dexter@cpan.org>

* Provides cpanfile.

0.0302  2023-06-10      Piotr Roszatycki <dexter@cpan.org>

* Default timeout changed to 60 seconds.
* Returns 504 if timeout is occured.

0.0301  2023-06-07      Piotr Roszatycki <dexter@cpan.org>

* New timeout option with default value 5 seconds.

0.0200  2016-10-29      Piotr Roszatycki <dexter@cpan.org>

* Non-blocking sockets.

0.0100  2014-11-16      Piotr Roszatycki <dexter@cpan.org>

* First CPAN release.


Dist.SH
#!/usr/bin/env bash

set -e

PERL=${PERL:-perl}
PROVE=${PROVE:-prove}

test -x Build && ./Build distclean

${PERL} Build.PL

rm -f MANIFEST

${PERL} -ni -e 'print unless /^## NAME/ .. eof' README.md
dist_version_from=$(${PERL} -ne 'print if /dist_version_from/' Build.PL | ${PERL} -pe "s%.*=> '%%; s%',%%")
${PERL} -MPod::Markdown -e '$p=Pod::Markdown->new;$p->parse_from_filehandle(\*STDIN);print $p->as_markdown' \
    <"${dist_version_from:-lib/$(grep ^name: MYMETA.yml | ${PERL} -pe 's/^name: //; s,-,/,g').pm}" |
    ${PERL} -pe 'sub a { $_[0]=~tr/ /-/; $_[0] } s,[(]http://search.cpan.org/perldoc[?]([^#)]*)(#[^)]*)?[)],a("(http://metacpan.org/module/$1$2)"),eg' |
    ${PERL} -pe 's/^# /## /' \
        >>README.md

${PERL} Build manifest

${PERL} Build distdir
diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST

cd ./*-[0-9]*.[0-9]*

${PERL} Build.PL
${PERL} Build

${PROVE} -Iblib/lib t

if [[ -z ${TEST_SKIP_EXTRA} ]] || [[ ${TEST_SKIP_EXTRA} == 0 ]]; then
    ${PROVE} -Iblib/lib xt
elif [[ -n ${TEST_SKIP_EXTRA} ]] && [[ ${TEST_SKIP_EXTRA} != 1 ]]; then
    # shellcheck disable=SC2046
    ${PROVE} $(find xt -name '*.t' ! -name "${TEST_SKIP_EXTRA}")
fi

if [[ $1 == sign ]]; then
    ${PERL} Build distsign
    diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST
else
    ${PERL} Build distdir
fi

chmod -R u+w,go-w ./*-[0-9]*.[0-9]*

export COPYFILE_DISABLE=true
tar zcf "$(basename "$(pwd)").tar.gz" ./*-[0-9]*.[0-9]*

mv ./*.tar.gz ..


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;
use Module::CPANfile;

my $module = 'Plack::Middleware::Proxy::Connect::IO';

(my $dist = $module) =~ s/::/-/g;

my $prereqs = Module::CPANfile->load->prereqs;

my $build = Module::Build->new(
    module_name             => $module,
    license                 => 'perl',
    build_requires          => $prereqs->requirements_for('build', 'requires')->as_string_hash,
    requires                => $prereqs->requirements_for('runtime', 'requires')->as_string_hash,
    recommends              => $prereqs->requirements_for('runtime', 'recommends')->as_string_hash,
    test_requires           => $prereqs->requirements_for('test', 'requires')->as_string_hash,
    auto_configure_requires => 0,
    create_makefile_pl      => 'traditional',
    create_readme           => 1,
    create_license          => 1,
    recursive_test_files    => 0,
    add_to_cleanup          => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://github.com/dex4er/perl-$dist/issues",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;


lib/Plack/Middleware/Proxy/Connect/IO.pm
package Plack::Middleware::Proxy::Connect::IO;

=head1 NAME

Plack::Middleware::Proxy::Connect::IO - CONNECT method

=head1 SYNOPSIS

=for markdown ```perl

    # In app.psgi
    use Plack::Builder;
    use Plack::App::Proxy;

    builder {
        enable "Proxy::Connect::IO", timeout => 30;
        enable "Proxy::Requests";
        Plack::App::Proxy->new->to_app;
    };

=for markdown ```

=head1 DESCRIPTION

This middleware handles the C<CONNECT> method. It allows to connect to
C<https> addresses.

The middleware runs on servers supporting C<psgix.io> and provides own
event loop so does not work correctly with C<psgi.nonblocking> servers.

The middleware uses only Perl's core modules: L<IO::Socket::INET> and
L<IO::Select>.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0304';

use parent qw(Plack::Middleware);

use Plack::Util::Accessor qw(
    timeout
);

use IO::Socket::INET;
use IO::Select;
use Socket qw(IPPROTO_TCP TCP_NODELAY);

use constant CHUNKSIZE       => 64 * 1024;
use constant DEFAULT_TIMEOUT => 60;
use constant READ_TIMEOUT    => 0.5;
use constant WRITE_TIMEOUT   => 0.5;

sub prepare_app {
    my ($self) = @_;

    # the default values
    $self->timeout(DEFAULT_TIMEOUT) unless defined $self->timeout;
}

sub call {
    my ($self, $env) = @_;

    return $self->app->($env)
        unless $env->{REQUEST_METHOD} eq 'CONNECT';

    return [501, [], ['']]
        unless $env->{'psgi.streaming'} and $env->{'psgix.io'};

    return sub {
        my ($respond) = @_;

        my $client = $env->{'psgix.io'};

        my ($host, $port) = $env->{REQUEST_URI} =~ m{^(?:.+\@)?(.+?)(?::(\d+))?$};

        my $remote = IO::Socket::INET->new(
            PeerAddr => $host,
            PeerPort => $port,
            Blocking => 0,
            Timeout  => $self->timeout,
        );

        if (!$remote) {
            if ($! eq 'Operation timed out') {
                return $respond->([504, [], ['']]);
            } else {
                return $respond->([502, [], ['']]);
            }
        }

        $client->blocking(0);

        # missing on Android
        if (eval { TCP_NODELAY }) {
            $client->setsockopt(IPPROTO_TCP, TCP_NODELAY, 1);
            $remote->setsockopt(IPPROTO_TCP, TCP_NODELAY, 1);
        }

        my $ioset = IO::Select->new;

        $ioset->add($client);
        $ioset->add($remote);

        my $writer = $respond->([200, []]);

        my $bufin = '';
        my $bufout = '';

    IOLOOP: while (1) {
            for my $socket ($ioset->can_read(READ_TIMEOUT)) {
                my $read = $socket->sysread(my $chunk, CHUNKSIZE);

                if ($read) {
                    if ($socket == $client) {
                        $bufout .= $chunk;
                    } elsif ($socket == $remote) {
                        $bufin .= $chunk;
                    }
                } else {
                    $client->syswrite($bufin);
                    $client->close;
                    $remote->syswrite($bufout);
                    $remote->close;
                    last IOLOOP;
                }
            }

            for my $socket ($ioset->can_write(WRITE_TIMEOUT)) {
                if ($socket == $client and length $bufin) {
                    my $write = $socket->syswrite($bufin);
                    substr $bufin, 0, $write, '';
                } elsif ($socket == $remote and length $bufout) {
                    my $write = $socket->syswrite($bufout);
                    substr $bufout, 0, $write, '';
                }
            }
        }
    };
}

1;

=head1 CONFIGURATION

=over 4

=item timeout

Timeout for the socket. The default value is C<60> seconds.

=back

=for readme continue

=head1 SEE ALSO

L<Plack>, L<Plack::App::Proxy>, L<Plack::Middleware::Proxy::Connect>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<https://github.com/dex4er/perl-Plack-Middleware-Proxy-Connect-IO/issues>

The code repository is available at
L<http://github.com/dex4er/perl-Plack-Middleware-Proxy-Connect-IO>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2014, 2016, 2023 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>
